File: unreal/AIArenaSDK/Source/AIArenaSDK/Private/Inference/InferenceGateway.cpp
#include "Inference/InferenceGateway.h"
#include "Inference/FallbackBehaviourAI.h"
#include "Core/AIArenaApiClient.h"
#include "Core/AIArenaConfig.h"
#include "Dom/JsonObject.h"
#include "Dom/JsonValue.h"
#include "TimerManager.h"
#include "Engine/World.h"

void UAIArenaInferenceGateway::GetCombatAction(
    const FString&           AgentId,
    const FAIArenaGameState& GameState,
    const TArray<FString>&   MemoryContext,
    FOnInferenceResult       OnResult)
{
    const UAIArenaConfig* Cfg = UAIArenaConfig::Get();
    const FString ReqId = FGuid::NewGuid().ToString();

    // ── Hard timeout: deliver FALLBACK if network takes too long ─────────────
    const float TimeoutSec = (float)Cfg->InferenceTimeoutMs / 1000.f;
    FTimerHandle Handle;
    GetWorld()->GetTimerManager().SetTimer(
        Handle,
        FTimerDelegate::CreateUObject(this, &UAIArenaInferenceGateway::OnRequestTimeout,
                                      ReqId, OnResult),
        TimeoutSec, /*bLoop=*/false);
    PendingTimers.Add(ReqId, Handle);

    // ── Build request body ────────────────────────────────────────────────────
    auto PosPart = MakeShared<FJsonObject>();
    PosPart->SetNumberField(TEXT("x"), GameState.Position.X);
    PosPart->SetNumberField(TEXT("y"), GameState.Position.Y);
    PosPart->SetNumberField(TEXT("z"), GameState.Position.Z);

    auto StatePart = MakeShared<FJsonObject>();
    StatePart->SetObjectField(TEXT("position"),      PosPart);
    StatePart->SetNumberField(TEXT("health"),        GameState.MaxHealth > 0.f
                                                       ? GameState.Health / GameState.MaxHealth
                                                       : 0.f);
    StatePart->SetNumberField(TEXT("timeRemaining"), GameState.TimeRemaining);
    StatePart->SetNumberField(TEXT("nearbyEnemies"), GameState.NearbyEnemies.Num());

    TArray<TSharedPtr<FJsonValue>> MemArray;
    for (const FString& Snippet : MemoryContext)
    {
        MemArray.Add(MakeShared<FJsonValueString>(Snippet));
    }

    auto Body = MakeShared<FJsonObject>();
    Body->SetStringField(TEXT("agentId"),       AgentId);
    Body->SetObjectField(TEXT("battleState"),   StatePart);
    Body->SetArrayField (TEXT("memoryContext"), MemArray);

    // ── Fire HTTP request ─────────────────────────────────────────────────────
    FAIArenaApiClient::Post(TEXT("/inference/action"), Body,
        FOnApiResponse::CreateLambda(
            [this, ReqId, OnResult](bool bOK, TSharedPtr<FJsonObject> Json)
            {
                // Cancel timeout timer if response arrived in time
                if (FTimerHandle* H = PendingTimers.Find(ReqId))
                {
                    if (GetWorld())
                        GetWorld()->GetTimerManager().ClearTimer(*H);
                    PendingTimers.Remove(ReqId);
                }
                else
                {
                    // Timeout already fired — discard this late response
                    return;
                }

                FAIArenaAgentAction Action;
                Action.Source = TEXT("AI");

                if (bOK && Json.IsValid())
                {
                    const TSharedPtr<FJsonObject>* ActionObj = nullptr;
                    if (Json->TryGetObjectField(TEXT("action"), ActionObj))
                    {
                        (*ActionObj)->TryGetStringField(TEXT("actionType"), Action.ActionType);
                        (*ActionObj)->TryGetNumberField(TEXT("confidence"), Action.Confidence);

                        FString Src;
                        if ((*ActionObj)->TryGetStringField(TEXT("source"), Src))
                            Action.Source = Src;
                    }
                }
                else
                {
                    Action.ActionType = TEXT("DEFEND");
                    Action.Confidence = 0.2f;
                    Action.Source     = TEXT("FALLBACK");
                }

                OnResult.ExecuteIfBound(Action);
            }));
}

void UAIArenaInferenceGateway::OnRequestTimeout(FString RequestId,
                                                  FOnInferenceResult OnResult)
{
    PendingTimers.Remove(RequestId);
    UE_LOG(LogTemp, Warning,
           TEXT("[InferenceGateway] Request %s timed out — using FALLBACK."), *RequestId);

    FAIArenaAgentAction Fallback;
    Fallback.ActionType = TEXT("DEFEND");
    Fallback.Confidence = 0.2f;
    Fallback.Source     = TEXT("FALLBACK");
    OnResult.ExecuteIfBound(Fallback);
}

File: unity/AIArenaSDK/Runtime/Inference/InferenceGateway.cs
using System;
using System.Threading.Tasks;
using UnityEngine;
using AIArena.SDK.Core;
using AIArena.SDK.Agents;

namespace AIArena.SDK.Inference
{
    /// <summary>
    /// HTTP client for the AI Arena inference-service.
    /// </summary>
    public class InferenceGateway : MonoBehaviour
    {
        /// <summary>Request a combat action from the inference API.</summary>
        public async Task<InferenceResult> GetCombatAction(string agentId, GameState gameState, string memoryContext = null)
        {
            var body = new
            {
                agentId,
                observationState = new
                {
                    position = new { x = gameState.Position.x, y = gameState.Position.y, z = gameState.Position.z },
                    health = gameState.Health / gameState.MaxHealth,
                    timeRemaining = gameState.TimeRemaining,
                    nearbyEnemies = gameState.NearbyEnemies?.Length ?? 0,
                },
                availableActions = gameState.AvailableActions,
                memoryContext,
            };

            var response = await ApiClient.Post<InferenceResponse>("/combat-action", body, AIArenaSDK.Instance.Config);
            return new InferenceResult
            {
                ActionType = response?.Action?.ActionType ?? "IDLE",
                Confidence = response?.Action?.Confidence ?? 0f,
                Source = response?.Action?.Source ?? "AI",
            };
        }
    }

    [Serializable]
    public class InferenceResponse
    {
        public InferenceActionData Action;
    }

    [Serializable]
    public class InferenceActionData
    {
        public string ActionType;
        public float Confidence;
        public string Source;
    }

    [Serializable]
    public class InferenceResult
    {
        public string ActionType;
        public float Confidence;
        public string Source;
    }
}
